testsuite/tests/typing-poly/poly.ml

```diff
@@ -1852,3 +1852,27 @@ Error: This expression has type < x : 'b. 'b s list >
        'a list
        The universal variable 'b would escape its scope
 |}]
+
+(* #9856 *)
+let f x =
+  let ref : type a . a option ref = ref None in
+  ref := Some x;
+  Option.get !ref
+[%%expect{|
+Line 2, characters 6-44:
+2 |   let ref : type a . a option ref = ref None in
+          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: This definition has type 'a option ref which is less general than
+         'a0. 'a0 option ref
+|}]
+
+type pr = { foo : 'a. 'a option ref }
+let x = { foo = ref None }
+[%%expect{|
+type pr = { foo : 'a. 'a option ref; }
+Line 2, characters 16-24:
+2 | let x = { foo = ref None }
+                    ^^^^^^^^
+Error: This field value has type 'b option ref which is less general than
+         'a. 'a option ref
+|}]
```

Comment:
Currently the regression test checks `type a . a option ref`, which is unsoundly accepted in trunk, and `{ foo : 'a . 'a option ref }`, which is correctly rejected. I would be in favor of adding two other tests to cover different forms of explicit polymorphism:

```ocaml
let test : 'a . 'a list ref = ref []  (* unsound in trunk *)
let test (type a) : a list ref = ref [] (* accepted in trunk, with a weak polymorphic type, which I find surprising *)
```

(Note: arguably `type a . a list ref` is a combination of both those features, but I prefer to (also) test the simpler forms. I didn't expect that the `(type a)` version would type-check, for example.)

Reply:
Locally abstract types are quite orthogonal to explicit polymorphism, and thus these two tests seem quite redundant too me?

Reply:
I had the impression that the use of a locally abstract type would guarantee that the definition would be (strongly) polymorphic under the corresponding type variable, so clearly the two features are not orthogonal *to me*. I have no idea whether it was just me thinking this, or whether this is a wide-spread assumption.

I think that my impression partly comes from the similarity between `let t (type a) ..` and the System F syntax for abstraction over a type variable. I read this code as if `a` was a name for a type (treated abstractly in the definition) that can be chosen by each *user* of `t`, sort of like a lightweight functor notation. With this interpretation the currently-accepted behavior `let foo (type a) : a list ref = []` is wrong. 

(Another way to object to this behavior is to say that a weak type variable is scoped *outside* the declaration that contains it, so accepting it to unify it with a locally abstract type could be treated as an escape of that local type.)

I just checked, and [the documentation](http://caml.inria.fr/pub/docs/manual-ocaml/firstclassmodules.html#s%3Afirst-class-modules) does not say that `(type a)` will be replaced by a a *generalizable* type variable, just a type variable, so the documentation is not contradicted by this example. I still find it surprising. (But this is orthogonal to the present bug.)

Reply:
Actually, the paragraph [Polymorphic syntax](http://caml.inria.fr/pub/docs/manual-ocaml/locallyabstract.html#p:polymorpic-locally-abstract) is quite explicit about this behavior:

> The `(type typeconstr-name)` syntax construction by itself does not make polymorphic the type variable it introduces, but it can be combined with explicit polymorphic annotations where needed.